storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/priority/GenericResourceAwareSchedulingPriorityStrategy.java

```diff
@@ -0,0 +1,178 @@
+         * The final score is a max over all resource types.
+         * Topology score will fall into the following intervals if:
+         *      User is under quota (guarantee):                    [(-guarantee)/available : 0]
+         *      User is over quota:                                 (0, infinity)
+         * Unfortunately, score below 0 does not guarantee that the topology will be scheduled due to resources fragmentation.
+         * @param availableCpu available CPU on the cluster.
+         * @param availableMemory available memory on the cluster.
+         * @param availableGenericResources available generic resources (other that cpu and memory) in cluster
+         * @param td the topology we are looking at.
+         * @return the score.
+         */
+        protected double getScore(double availableCpu, double availableMemory,
+                                  Map<String, Double> availableGenericResources, TopologyDetails td) {
+            // calculate scores for cpu and memory first
+            double ret = super.getScore(availableCpu, availableMemory, td);
+            if (ret == Double.MAX_VALUE) {
+                return ret;
+            }
+            Map<String, Double> tdTotalRequestedGeneric = td.getTotalRequestedGenericResources();
+            if (tdTotalRequestedGeneric == null) {
+                tdTotalRequestedGeneric = Collections.emptyMap();
+            }
+            for (Map.Entry<String, Double> entry : availableGenericResources.entrySet()) {
+                String resource = entry.getKey();
+                Double available = entry.getValue();
+                if (available <= 0) {
+                    return Double.MAX_VALUE;
+                }
+                Double wouldBeResource = assignedGenericResources.getOrDefault(resource, 0.0)
+                    + tdTotalRequestedGeneric.getOrDefault(resource, 0.0);
+                double thisScore = (wouldBeResource -  guaranteedGenericResources.getOrDefault(resource, 0.0)) / available;
+                ret = Math.max(ret, thisScore);
+            }
+
+            return ret;
+        }
+
+        protected double getScore(double availableCpu, double availableMemory, Map<String, Double> availableGenericResources) {
+            TopologyDetails td = getNextHighest();
+            return getScore(availableCpu, availableMemory, availableGenericResources, td);
+        }
+    }
+
+    private static class GrasSimulatedUserComparator implements Comparator<GrasSimulatedUser> {
+        private final double cpuAvail;
+        private final double memAvail;
+        private final Map<String, Double> genericAvail;
+
+        private GrasSimulatedUserComparator(double cpuAvail, double memAvail, Map<String, Double> genericAvail) {
+            this.cpuAvail = cpuAvail;
+            this.memAvail = memAvail;
+            this.genericAvail = genericAvail;
+        }
```

Comment:
I think this can be removed. Equivalent to Comparator<GrasSimulatedUser> c = (o1,o2) -> Double.compare(o1.getScore(cpuAvail, memAvail, genericAvail), o2.getScore(cpuAvail, memAvail, genericAvail));